### DISCUSSION TITLE
USB Stall when using printf

### DISCUSSION THREAD

**Discussion by UofTPhyEssSW** [earlephilhower/arduino-pico#1939] [Q&A]
I have a custom RP2040 board in the Arduino UNO Rev3 form factor. I have written an example sketch to read a 4-Channel SPI ADC (MAX11635) on the board. This sketch freezes up when the printf() statements are uncommented. When the printf() statement commented out the sketch runs fine with no freezing up. I have confirmed that the MAX11365 library is working and not the cause of this problem. 

Here is the sketch I am using.

```cpp
#include "max11635.h"

#define ARDUINO_SERIAL_PLOTTER  // Comment out for text output instead of Serial plotter format.

// Phyduino Pico Rev D GPIO assignments (Change according to board being used.)
static constexpr std::uint32_t LED_IO       { 5 };
static constexpr std::uint32_t ADC_MOSI_IO  { 23 };
static constexpr std::uint32_t ADC_MISO_IO  { 20 };
static constexpr std::uint32_t ADC_SCK_IO   { 22 };
static constexpr std::uint32_t ADC_CS_IO    { 21 };
static constexpr std::uint32_t ADC_nCNST_IO { 18 };
static constexpr std::uint32_t ADC_nEOC_IO  { 19 };

// Sample Buffer
max11635::driver::data_type Sample [max11635::driver::max_channels] { };

void setup() {
  Serial.begin();
  // Set board LED initialize state
  pinMode(LED_IO, OUTPUT);
  digitalWrite(LED_IO, LOW);
  // Configure MAX11635 ADC GPIO's, must be called before begin()!
  MAX11635_ADC.configure_io(ADC_MOSI_IO, ADC_MISO_IO, ADC_SCK_IO, ADC_CS_IO, ADC_nCNST_IO, ADC_nEOC_IO);
  // Initialize MAX11635 ADC (Default SPI clock frequency 4.8MHz)
  MAX11635_ADC.begin();
}

void loop() {
  delay(200);
  digitalWrite(LED_IO, HIGH);

  // Read channels
  for(std::size_t i = 0; i < max11635::driver::max_channels; i++){
    Sample[i] = MAX11635_ADC.analogRead(i);

    #ifdef ARDUINO_SERIAL_PLOTTER
    if(i == max11635::driver::max_channels - 1){
      Serial.printf("CH[%d]:%.3f\r\n", i, max11635::driver::to_voltage(Sample[i]));
    } else{
      Serial.printf("CH[%d]:%.3f,", i, max11635::driver::to_voltage(Sample[i]));
    }
    #else
    // Read Analog value
    Serial.printf("Channel[%d] : %.3fV [0x%04x]\r\n", i, max11635::driver::to_voltage(Sample[i]), Sample[i]);
    #endif
  }

  delay(100);
  digitalWrite(LED_IO, LOW);
}
```
The max11635 library can be found here (https://github.com/UofTPhyEssSW/MAX11635-ADC). 

## Device Settings
- Board "Generic RP2040"
- Boot Stage 2 :  "W25Q128JV QSPI/4"
- Debug Level : "None"
- Debug Port: "Disabled"
- C++ Exceptions: "Disabled"
- Flash Size : "16MB NO FS"
- CPU Speed: "133MHz"
- IP/Bluetooth Stack: "UPv4 Only"
- Optimize : "Small (-Os) (standard)"
- RTTI : "Disabled"
- Stack Protector: "Disabled"
- Upload Method : "Default (UF2)"
- USB Stack : "Pico SDK"

From what I can tell this looks like a USB TX buffer overrun issue and I am pushing to much data to the USB buffer. Other tests that I have done have only counted number of samples taken and only output the count every 400 samples (~10sec). Those tests have run over night with no issues. 

Any Insight to this issue would be helpful. 

Thanks
Rob Morley.

**Comment by earlephilhower** [comment]
The USB serial TX buffer can't overrun, but it can block if the computer doesn't eat the USB packets faster than the Pico produces them.  You can see this with the 22khz x 16bit x 2channel `I2SInput` example which also uses the serial plotter.  And if you try and write to `Serial` from an interrupt that could be bad as well since you could already be trying to write to Serial in the main app and you should never block in an IRQ.

If you can make a small MCVE that reproduces the hang w/o any special hardware others can look at it, but if not then you're pretty much on your own since nobody else will have your specific HW combo and 3rd party library.  Your best bet then is to hook up a PicoProbe and dump the stack to see where the core is when it "hangs" (also I'm not sure what you mean by hang, specifically...).
